### ISSUE TITLE
Fix S2190 FP/FN: "Add a way to break out of this method's recursion"

### ISSUE THREAD

**Issue by noseratio** [SonarSource/sonar-dotnet#7364]
### Description

Breaking out from `while(true)` loops via cancellation token / `OperaitionCancelledException` should not be flagged as a "Blocker".

### Repro steps

```C#
 public async IAsyncEnumerable<int> GenerateSequenceAsync(
        [EnumeratorCancellation] CancellationToken cancelToken)
    {
        int i = 0;
        while (true)
        {
            await Task.Delay(100, cancelToken);
            yield return i++;
        }
    }
```

### Expected behavior

Altering the flow of code execution via .NET cancellation protocol might deserve a warning, but I disagree it's a blocker. 

[More detailed reasoning](https://www.reddit.com/r/dotnet/comments/145k670/comment/jnljnsp/?utm_source=share&utm_medium=web2x&context=3). 

### Actual behavior

Sonar reports a blocker: 

> Add a way to break out of this method's recursion. Recursion should not be infinite [csharpsquid:S2190](https://sonarcloud.io/organizations/emdat/rules?open=csharpsquid%3AS2190&rule_key=csharpsquid%3AS2190)

### Known workarounds

```C#
 public async IAsyncEnumerable<int> GenerateSequenceAsync(
        [EnumeratorCancellation] CancellationToken cancelToken)
    {
        int i = 0;
        while (true)
        {
            if (cancelToken.IsCancellationRequested)
            {
                // deal with "a-way-to-break-out-of-here" warning
                throw new TaskCanceledException(null, null, cancelToken);
            }
            await Task.Delay(100, cancelToken);
            yield return i++;
        }
    }
```

**Comment by gregory-paidis-sonarsource** [comment]
Hey @noseratio , thanks for raising this, it's a really good catch.

I added it to our backlog, where it will be taken care of during the next hardening sprint.

From my point of view, we should handle two distinct cases as valid exit conditions:

- when `token.ThrowIfCancellationRequested` is called 
- when a `CancellationToken` is passed down to an `async` method

Please inform me if I missed anything, and thanks again for the suggestion!

**Comment by noseratio** [comment]
Hey @gregory-paidis-sonarsource ,

Thanks for accepting it!

> From my point of view, we should handle two distinct cases as valid exit conditions:
> 
> * when `token.ThrowIfCancellationRequested` is called
> * when a `CancellationToken` is passed down to an `async` method
> 
> Please inform me if I missed anything, and thanks again for the suggestion!

IMO, this should cover it well! If you'd like to go deeper, there could be a bit more involved cases, like creating a linked child token and passing that one down 🙂

For example:

```
 public async IAsyncEnumerable<int> GenerateSequenceAsync(
        [EnumeratorCancellation] CancellationToken cancelToken)
    {
        using var childCts = CancellationTokenSource.CreateLinkedTokenSource(cancelToken).
        childCts.CancelAfter(TimeSpan.FromSeconds(1));

        int i = 0;
        while (true)
        {
            await Task.Delay(100, childCts.Token);
            yield return i++;
        }
    }
```

**Comment by pavel-mikula-sonarsource** [comment]
Thanks @fiotti for this snippet. 

I'd say `while (!cancelToken.IsCancellationRequested)` is not a workaround, but a proper solution to this situation. We could just document it that way and close this ticket.

Or maybe even have a dedicated new rule idea about the cancellation of infinite loops based on this. @gregory-paidis-sonarsource, WDYT?

**Comment by gregory-paidis-sonarsource** [comment]
@fiotti Can you elaborate on why it would not work for `Task`?

For example, this works as expected:
```csharp
    public static async Task<List<int>> GenerateSequenceAsync(CancellationToken cancelToken)
    {
        int i = 0;
        var res = new List<int>();
        while (!cancelToken.IsCancellationRequested)
        {
            await Task.Delay(100, CancellationToken.None); // Or cancelToken, if you need to throw here
            res.Add(i);
            i++;
        }
        return res;
    }
```
---
@pavel-mikula-sonarsource  I agree with your opinion. 
I would not make a new rule, but I do think it is worth documenting.
Do you think we should explicitly mention it on the rule's RSPEC as an edge-case, or just add some UTs around this?

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1770](https://sonarsource.atlassian.net/browse/NET-1770)

[NET-1770]: https://sonarsource.atlassian.net/browse/NET-1770?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by denisbredikhin** [comment]
Hello @pavel-mikula-sonarsource @gregory-paidis-sonarsource , 
I want to comment on this. 
Using `while (!cancelToken.IsCancellationRequested)` is not a proper solution for this case.
Caller may expect that the code throw an exception when cancellation is requested, to know if it was respected or not.
Silent exit from the method in case of cancellation is an anti-pattern.
See more information here: https://blog.stephencleary.com/2022/03/cancellation-4-polling.html
So can you please add the exception to S2190 in case when `ThrowIfCancellationRequested` is called?
